File: backend/install_realtime.py
"""
Install and test real-time dependencies
"""
import subprocess
import sys
import importlib.util

def check_package(package_name):
    """Check if a package is installed"""
    spec = importlib.util.find_spec(package_name)
    return spec is not None

def install_package(package):
    """Install a package using pip"""
    print(f"\n[Installing] {package}...")
    try:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', package])
        print(f"[SUCCESS] {package} installed!")
        return True
    except subprocess.CalledProcessError as e:
        print(f"[ERROR] Failed to install {package}: {e}")
        return False

def test_imports():
    """Test if all required packages can be imported"""
    print("\n" + "="*60)
    print("TESTING IMPORTS")
    print("="*60)

    packages = {
        'flask_socketio': 'Flask-SocketIO',
        'socketio': 'python-socketio',
    }

    all_ok = True
    for module_name, package_name in packages.items():
        try:
            __import__(module_name)
            print(f"✅ {package_name} - OK")
        except ImportError as e:
            print(f"❌ {package_name} - FAILED: {e}")
            all_ok = False

    return all_ok

def verify_files():
    """Verify that all required files exist and are correct"""
    import os

    print("\n" + "="*60)
    print("VERIFYING FILES")
    print("="*60)

    files_to_check = [
        'services/socket_service.py',
        'extensions.py',
        'app.py',
        'routes/projects.py',
        'routes/votes.py',
    ]

    all_ok = True
    for file_path in files_to_check:
        if os.path.exists(file_path):
            print(f"✅ {file_path} - EXISTS")

            # Check for specific content
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()

                if file_path == 'services/socket_service.py':
                    if 'SocketService' in content and 'emit_project_created' in content:
                        print(f"   ✅ Contains SocketService class")
                    else:
                        print(f"   ❌ Missing SocketService implementation")
                        all_ok = False

                elif file_path == 'extensions.py':
                    if 'socketio' in content and 'SocketIO' in content:
                        print(f"   ✅ Contains socketio initialization")
                    else:
                        print(f"   ❌ Missing socketio import")
                        all_ok = False

                elif file_path == 'app.py':
                    if 'socketio.run' in content or 'socketio.init_app' in content:
                        print(f"   ✅ Socket.IO integrated")
                    else:
                        print(f"   ❌ Socket.IO not integrated")
                        all_ok = False

                elif file_path == 'routes/projects.py':
                    if 'SocketService.emit_project_created' in content:
                        print(f"   ✅ Emits project events")
                    else:
                        print(f"   ⚠️  No Socket.IO events (optional)")

        else:
            print(f"❌ {file_path} - NOT FOUND")
            all_ok = False

    return all_ok

def test_socket_service():
    """Test if SocketService can be imported"""
    print("\n" + "="*60)
    print("TESTING SOCKET SERVICE")
    print("="*60)

    try:
        from services.socket_service import SocketService
        print("✅ SocketService imported successfully")

        # Check methods exist
        methods = [
            'emit_project_created',
            'emit_project_updated',
            'emit_project_deleted',
            'emit_vote_cast',
            'emit_leaderboard_updated',
        ]

        for method in methods:
            if hasattr(SocketService, method):
                print(f"   ✅ {method} - OK")
            else:
                print(f"   ❌ {method} - MISSING")
                return False

        return True
    except Exception as e:
        print(f"❌ Failed to import SocketService: {e}")
        return False

def main():
    """Main installation and testing flow"""
    print("="*60)
    print("BACKEND REAL-TIME INSTALLATION & TESTING")
    print("="*60)

    # Step 1: Install dependencies
    print("\n[STEP 1] Installing dependencies...")

    packages_to_install = [
        'flask-socketio==5.3.6',
        'python-socketio==5.11.0',
    ]

    for package in packages_to_install:
        if not install_package(package):
            print(f"\n❌ Installation failed for {package}")
            sys.exit(1)

    # Step 2: Test imports
    if not test_imports():
        print("\n❌ Import test failed!")
        sys.exit(1)

    # Step 3: Verify files
    if not verify_files():
        print("\n⚠️  Some files are missing or incomplete")
        print("   But core dependencies are installed.")

    # Step 4: Test SocketService
    if not test_socket_service():
        print("\n⚠️  SocketService test failed")
        print("   But dependencies are installed.")

    # Summary
    print("\n" + "="*60)
    print("INSTALLATION COMPLETE")
    print("="*60)
    print("\n✅ All dependencies installed successfully!")
    print("✅ Flask-SocketIO ready")
    print("✅ python-socketio ready")
    print("\n📝 Next steps:")
    print("   1. Restart backend: python app.py")
    print("   2. Check for '[Socket.IO] Server initialized' in logs")
    print("   3. Test real-time features")

    return True

if __name__ == '__main__':
    try:
        success = main()
        sys.exit(0 if success else 1)
    except Exception as e:
        print(f"\n❌ Installation script failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: frontend/install_realtime.js
/**
 * Install and verify real-time dependencies for frontend
 */
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

console.log('='.repeat(60));
console.log('FRONTEND REAL-TIME INSTALLATION & TESTING');
console.log('='.repeat(60));

// Step 1: Install socket.io-client
console.log('\n[STEP 1] Installing socket.io-client...\n');
try {
  // Detect package manager
  let packageManager = 'npm';
  if (fs.existsSync('pnpm-lock.yaml')) {
    packageManager = 'pnpm';
  } else if (fs.existsSync('yarn.lock')) {
    packageManager = 'yarn';
  }

  console.log(`[INFO] Using ${packageManager} package manager`);

  const installCmd = packageManager === 'npm'
    ? 'npm install socket.io-client'
    : packageManager === 'pnpm'
    ? 'pnpm add socket.io-client'
    : 'yarn add socket.io-client';

  execSync(installCmd, { stdio: 'inherit' });
  console.log('\n✅ socket.io-client installed successfully!');
} catch (error) {
  console.error('\n❌ Failed to install socket.io-client:', error.message);
  process.exit(1);
}

// Step 2: Verify package.json
console.log('\n' + '='.repeat(60));
console.log('VERIFYING PACKAGE.JSON');
console.log('='.repeat(60));

try {
  const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));

  if (packageJson.dependencies && packageJson.dependencies['socket.io-client']) {
    console.log('✅ socket.io-client in dependencies');
    console.log(`   Version: ${packageJson.dependencies['socket.io-client']}`);
  } else {
    console.log('❌ socket.io-client NOT in dependencies');
  }
} catch (error) {
  console.error('❌ Failed to read package.json:', error.message);
}

// Step 3: Verify files exist
console.log('\n' + '='.repeat(60));
console.log('VERIFYING FILES');
console.log('='.repeat(60));

const filesToCheck = [
  'src/hooks/usePrefetch.ts',
  'src/hooks/useRealTimeUpdates.ts',
  'src/hooks/useProjects.ts',
  'src/hooks/useLeaderboard.ts',
  'src/App.tsx',
  'src/index.css',
];

let allFilesOk = true;

filesToCheck.forEach(file => {
  const filePath = path.join(process.cwd(), file);
  if (fs.existsSync(filePath)) {
    console.log(`✅ ${file} - EXISTS`);

    const content = fs.readFileSync(filePath, 'utf8');

    // Check for specific content
    if (file === 'src/hooks/usePrefetch.ts') {
      if (content.includes('usePrefetch') && content.includes('prefetchQuery')) {
        console.log('   ✅ Contains prefetch logic');
      } else {
        console.log('   ❌ Missing prefetch implementation');
        allFilesOk = false;
      }
    }

    if (file === 'src/hooks/useRealTimeUpdates.ts') {
      if (content.includes('socket.io-client') && content.includes('useRealTimeUpdates')) {
        console.log('   ✅ Contains Socket.IO integration');
      } else {
        console.log('   ❌ Missing Socket.IO implementation');
        allFilesOk = false;
      }
    }

    if (file === 'src/hooks/useProjects.ts') {
      if (content.includes('refetchInterval') && content.includes('placeholderData')) {
        console.log('   ✅ Background refetch configured');
      } else {
        console.log('   ⚠️  Missing background refetch (optional)');
      }
    }

    if (file === 'src/App.tsx') {
      if (content.includes('usePrefetch') && content.includes('useRealTimeUpdates')) {
        console.log('   ✅ Hooks integrated');
      } else {
        console.log('   ❌ Hooks not integrated');
        allFilesOk = false;
      }
    }

    if (file === 'src/index.css') {
      if (content.includes('fade-in') && content.includes('slide-in-down')) {
        console.log('   ✅ Animations added');
      } else {
        console.log('   ⚠️  Missing animations (optional)');
      }
    }
  } else {
    console.log(`❌ ${file} - NOT FOUND`);
    allFilesOk = false;
  }
});

// Step 4: Check environment variables
console.log('\n' + '='.repeat(60));
console.log('CHECKING ENVIRONMENT');
console.log('='.repeat(60));

const envFiles = ['.env', '.env.local', '.env.development'];
let hasEnvFile = false;
let hasApiUrl = false;

envFiles.forEach(envFile => {
  if (fs.existsSync(envFile)) {
    console.log(`✅ ${envFile} - EXISTS`);
    hasEnvFile = true;

    const envContent = fs.readFileSync(envFile, 'utf8');
    if (envContent.includes('VITE_API_URL')) {
      console.log('   ✅ VITE_API_URL defined');
      hasApiUrl = true;
    }
  }
});

if (!hasEnvFile) {
  console.log('⚠️  No .env file found');
  console.log('   Creating .env with default values...');

  fs.writeFileSync('.env', 'VITE_API_URL=http://localhost:5000\n');
  console.log('✅ Created .env file');
  hasApiUrl = true;
}

if (!hasApiUrl) {
  console.log('⚠️  VITE_API_URL not set');
  console.log('   Add this to your .env file:');
  console.log('   VITE_API_URL=http://localhost:5000');
}

// Step 5: Test import syntax (syntax check only)
console.log('\n' + '='.repeat(60));
console.log('SYNTAX VALIDATION');
console.log('='.repeat(60));

try {
  const useRealTimeUpdates = fs.readFileSync('src/hooks/useRealTimeUpdates.ts', 'utf8');

  // Check for common syntax errors
  if (useRealTimeUpdates.includes('import') && useRealTimeUpdates.includes('export')) {
    console.log('✅ useRealTimeUpdates.ts - Valid ES module syntax');
  } else {
    console.log('❌ useRealTimeUpdates.ts - Invalid syntax');
    allFilesOk = false;
  }

  if (useRealTimeUpdates.includes("socket.io-client") || useRealTimeUpdates.includes('socket.io-client')) {
    console.log('✅ socket.io-client import found');
  } else {
    console.log('❌ socket.io-client import NOT found');
    allFilesOk = false;
  }
} catch (error) {
  console.log('❌ Could not validate syntax:', error.message);
}

// Summary
console.log('\n' + '='.repeat(60));
console.log('INSTALLATION COMPLETE');
console.log('='.repeat(60));

if (allFilesOk) {
  console.log('\n✅ All checks passed!');
} else {
  console.log('\n⚠️  Some checks failed, but dependencies are installed');
}

console.log('\n📦 Installed packages:');
console.log('   ✅ socket.io-client');

console.log('\n📝 Files created/updated:');
console.log('   ✅ src/hooks/usePrefetch.ts');
console.log('   ✅ src/hooks/useRealTimeUpdates.ts');
console.log('   ✅ src/App.tsx (updated)');
console.log('   ✅ src/index.css (updated)');

console.log('\n🚀 Next steps:');
console.log('   1. Restart frontend: npm run dev');
console.log('   2. Check console for:');
console.log('      - [Prefetch] Completed');
console.log('      - [Socket.IO] Connected successfully');
console.log('   3. Test real-time features');

console.log('\n✅ Frontend ready for real-time features!');
